libcontainer/nsenter/cloned_binary.c

```diff
@@ -415,29 +423,89 @@ static int try_bindfd(void)
 	close(fd);
 
 	/*
-	 * For obvious reasons this won't work in rootless mode because we haven't
-	 * created a userns+mntns -- but getting that to work will be a bit
-	 * complicated and it's only worth doing if someone actually needs it.
+	 * Daemons such as systemd and udisks2 watch /proc/self/mountinfo and
+	 * re-parse it on every change, which gets expensive when the mount table
+	 * is large and/or changes frequently. Set up the bind-mounts in a new,
+	 * private mount namespace so as not to wake up those processes every
+	 * time we nsexec into a container. We clone a child process into a new
+	 * mount namespace to do the dirty work so the side effects of unsharing
+	 * the mount namespace do not leak into the current process.
 	 */
-	ret = -EPERM;
-	if (mount("/proc/self/exe", template, "", MS_BIND, "") < 0)
+	if ((ret = socketpair(AF_LOCAL, SOCK_STREAM, 0, pipe)) < 0)
 		goto out;
-	if (mount("", template, "", MS_REMOUNT | MS_BIND | MS_RDONLY, "") < 0)
-		goto out_umount;
 
-	/* Get read-only handle that we're sure can't be made read-write. */
-	ret = open(template, O_PATH | O_CLOEXEC);
+	is_child = setjmp(env);
+	if (is_child) {
+		close(pipe[0]);
+		fd = -1;
+		/*
+		 * For obvious reasons this won't work in rootless mode because we haven't
+		 * created a userns -- but getting that to work will be a bit complicated and
+		 * it's only worth doing if someone actually needs it.
+		 */
+		if ((ret = mount("none", "/", NULL, MS_PRIVATE | MS_REC, NULL)) < 0)
+			goto out_child;
+		/*
+		 * The kernel refuses to bind-mount from the magic symlink when the process has
+		 * been cloned (or unshared) into a new mount namespace, at least on Linux 4.4.
+		 */
+		char linkbuf[PATH_MAX + 1] = { 0 };
+		ssize_t linkpathlen = readlink("/proc/self/exe", linkbuf, sizeof(linkbuf));
+		if (linkpathlen == sizeof(linkbuf)) {
+			/*
+			 * The link path is longer than PATH_MAX, and the contents of
+			 * linkbuf might have been truncated. A truncated path could
+			 * happen to be a valid path to a different file, which could
+			 * allow for local privilege escalation if we were to exec it.
+			 * The mount syscall doesn't accept paths longer than PATH_MAX,
+			 * anyway.
+			 */
+			ret = -ENAMETOOLONG;
+			goto out_child;
+		}
+		linkbuf[linkpathlen] = '\0';
```

Comment:
From the code it looks like this is redundant, since linkbuf is initialized with zeroes, and the `if` above guarantees that at least one zero remains in it. OTOH it doesn't hurt either.